examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,693 @@
+# strong negative impact on the predicted number of bike rentals.
+#
+# Time-steps as categories
+# ------------------------
+#
+# Since the time features are encoded in a discrete manner using integers (24
+# unique values in the "hours" feature), we could decide to treat those as
+# categorical variables and ignore any assumption implied by the ordering of
+# the hour values using a one-hot encoding.
+#
+# Using one-hot encoding for the time features gives the linear model a lot
+# more flexibility as we introduce one additional feature per discrete time
+# level.
+one_hot_linear_pipeline = make_pipeline(
+    ColumnTransformer(
+        transformers=[
+            ("categorical", one_hot_encoder, categorical_columns),
+            ("one_hot_time", one_hot_encoder, ["hour", "weekday", "month"]),
+        ],
+        remainder="passthrough",
+    ),
+    RidgeCV(alphas=alphas),
+)
+
+evaluate(one_hot_linear_pipeline, X, y, cv=ts_cv)
+
+# %%
+# The average error rate of this model is 10% which is much better than using
+# the original ordinal encoding of the time feature, confirming our intuition
+# that the linear regression model benefit from the added flexibility to not
+# treat time progression in a monotonic manner.
+#
+# However, this introduces a very large number of new features. If the time of
+# the day was represented in minutes since the start of the day instead of
+# hours, one-hot encoding would have introduced 1440 features instead of 24.
+# This could cause some significant overfitting. To avoid this we could use
+# :func:`sklearn.preprocessing.KBinsDiscretizer` instead to re-bin the number
+# of levels of fine-grained ordinal or numerical variables while still
+# benefitting from the non-monotonic expressivity advantages of one-hot
+# encoding.
+#
+# Finally, we also observe than one-hot encoding completely ignores the
+# ordering of the hour levels while this could be an interesting inductive bias
+# to preserve to some level. In the following we try to explore smooth, non
+# monotonic encoding that locally preserve the relative ordering of time
+# features.
+#
+# Trigonometric features
+# ----------------------
+#
+# As a first attempt, we can try to encode each of those periodic features
+# using a sine and cosine transform with the matching period.
+#
+# Each ordinal time feature is transformed into 2 features that together encode
+# equivalent information in a non-monotonic way, and more importantly without
+# any jump between the first and the last value of the periodic range.
+from sklearn.preprocessing import FunctionTransformer
+
+
+def sin_transformer(period):
+    return FunctionTransformer(lambda x: np.sin(x / period * 2 * np.pi))
+
+
+def cos_transformer(period):
+    return FunctionTransformer(lambda x: np.cos(x / period * 2 * np.pi))
+
+
+# %%
+#
+# Let us visualize the effect of this feature expansion on some synthetic hour
+# data with a bit of extrapolation beyond hour=23:
```

Comment:
Is it worth mentioning or showing that a parametric plot, i.e. the 2 features on x and y axis parameterized by the hour, gives a circle?

Reply:
I am not too sure how to do such an intuitive to interpret plot without too much matplotlib boilerplate. Maybe with color coded hour dots and a color bar? Or text annotations on the circule in the x/y space?

Reply:
```python
plt.scatter(hour_df["hour_sin"], hour_df["hour_cos"], c=hour_df["hour"])
```
This is for fun only. A circle is like an analog clock:laughing:

Reply:
Done.